src/builder/QuarkBuilder.sol

```diff
@@ -266,53 +266,93 @@ contract QuarkBuilder {
+                }
             }
 
-            if (tokenUsed == actions[i].paymentToken) {
+            plannedBridgeAmount += bridgeActionContext.amount;
+        }
+
+        // Verify transfer actions is affordable
+        uint256 accruedCost = 0;
+        for (uint256 i = 0; i < transferActions.length; ++i) {
+            Actions.TransferActionContext memory transferActionContext =
+                abi.decode(transferActions[i].actionContext, (Actions.TransferActionContext));
+            address tokenUsed = transferActionContext.token;
+            uint256 amountUsed = transferActionContext.amount;
+            accruedCost += transferActions[i].paymentMaxCost;
+            if (tokenUsed == transferActions[i].paymentToken) {
                 // If the payment token is the transfer token and this is the target chain, we need to account for the transfer amount
-                uint256 paymentAssetBalanceOnChain =
-                    Accounts.sumBalances(Accounts.findAssetPositions(tokenUsed, actions[i].chainId, chainAccountsList));
+                uint256 paymentAssetBalanceOnChain = Accounts.sumBalances(
+                    Accounts.findAssetPositions(tokenUsed, transferActions[i].chainId, chainAccountsList)
+                );
 
                 // If its transfer step, check if user has enough balance to cover the transfer amount after bridge
-                if (Strings.stringEqIgnoreCase(actions[i].actionType, Actions.ACTION_TYPE_TRANSFER)) {
-                    if (paymentAssetBalanceOnChain + plannedBridgeAmount < actions[i].paymentMaxCost + amountUsed) {
-                        revert MaxCostTooHigh();
-                    }
-                } else {
-                    if (paymentAssetBalanceOnChain < actions[i].paymentMaxCost + amountUsed) {
-                        revert MaxCostTooHigh();
-                    }
+                if (paymentAssetBalanceOnChain + plannedBridgeAmount < accruedCost + amountUsed) {
+                    revert MaxCostTooHigh();
                 }
+
+                // Special handling as the payment token is sent out so it will be part of the cost
+                accruedCost += amountUsed;
             } else {
                 // Just check payment token can cover the max cost
                 uint256 paymentAssetBalanceOnChain = Accounts.sumBalances(
-                    Accounts.findAssetPositions(actions[i].paymentToken, actions[i].chainId, chainAccountsList)
+                    Accounts.findAssetPositions(
+                        transferActions[i].paymentToken, transferActions[i].chainId, chainAccountsList
+                    )
                 );
 
-                if (paymentAssetBalanceOnChain < actions[i].paymentMaxCost) {
+                if (paymentAssetBalanceOnChain < accruedCost) {
                     revert MaxCostTooHigh();
                 }
             }
         }
     }
 
+    function findActionsOfType(Actions.Action[] memory actions, string memory actionType)
```

Comment:
This should live in the Actions library. I guess `truncate` for Actions should also be moved there.

Reply:
Sure I can move those existing functions to there as well.